File: Website/Agricultural Management Website Design (1)/src/app/lib/repairJobs.ts
import type { EquipmentConditionReport, IrrigationDamageReportRecord, WorkerRecord } from '../types/appState';
import { isMaintenanceRole } from './farmFinance';
import { isWorkerActive } from './workerUi';
import type { PartUsed } from './repairParts';

/**
 * A repair job: the admin sends a broken-equipment or sprinkler-damage report to ONE Maintenance worker.
 * Workers' reports live in each worker's private document, which other workers can't read, so the job
 * carries a copy of what Maintenance needs to see (see the `repair_jobs` rules in firestore.rules).
 * Kinds are open-ended: adding another report type later only needs another builder here.
 */
export type RepairKind = 'equipment' | 'sprinkler';
export type RepairStatus = 'assigned' | 'in_progress' | 'fixed';

export interface RepairJobRecord {
  kind: RepairKind;
  reportId: string;
  /** What Maintenance sees: the equipment name or the sprinkler label, plus a short context line. */
  title: string;
  subtitle: string;
  details: string;
  reportedBy: string;
  reportedAt: string;
  assignedToUid: string;
  assignedToName: string;
  assignedToWorkerId?: string;
  assignedAt: string;
  status: RepairStatus;
  /** Written by the Maintenance worker's app. */
  startedAt?: string | null;
  fixedAt?: string | null;
  /** What was done, required when marking the job fixed. */
  fixNote?: string | null;
  /** Parts / materials the worker used (written with fixNote); the admin deducts them from the supplies. */
  partsUsed?: PartUsed[] | null;
  /** Set by the admin once partsUsed were taken out of the supplies, so it can only happen once. */
  partsDeductedAt?: string | null;
}

export interface Assignee {
  uid: string;
  name: string;
  workerId: string;
}

export const repairJobId = (kind: RepairKind, reportId: string) => `${kind}-${reportId}`;

/** Active Maintenance workers who can actually sign in to the app (a job is matched to them by login uid). */
export function maintenanceAssignees(workers: WorkerRecord[]): Assignee[] {
  return workers
    .filter((w) => isMaintenanceRole(w.roleRate) && isWorkerActive(w) && Boolean(w.authUid?.trim()))
    .map((w) => ({ uid: w.authUid!.trim(), name: w.name, workerId: w.workerId?.trim() ?? '' }));
}

export function equipmentRepairJob(report: EquipmentConditionReport, to: Assignee, now = new Date().toISOString()): RepairJobRecord {
  return {
    kind: 'equipment',
    reportId: report.reportId,
    title: report.equipmentName || 'Equipment',
    subtitle: report.isWrecked ? 'Broken / wrecked' : 'Needs maintenance',
    details: report.notes?.trim() || 'No details were given.',
    reportedBy: report.reportedBy?.trim() || 'Worker',
    reportedAt: report.reportedAt,
    assignedToUid: to.uid,
    assignedToName: to.name,
    assignedToWorkerId: to.workerId,
    assignedAt: now,
    status: 'assigned',
    startedAt: null,
    fixedAt: null,
    fixNote: null,
  };
}

export function sprinklerRepairJob(report: IrrigationDamageReportRecord, to: Assignee, now = new Date().toISOString()): RepairJobRecord {
  const label = report.sprinklerLabel?.trim() || report.zone?.trim() || 'Sprinkler';
  return {
    kind: 'sprinkler',
    reportId: report.reportId ?? '',
    title: label,
    subtitle: report.zone?.trim() && report.zone.trim() !== label ? `Zone: ${report.zone.trim()}` : 'Sprinkler damage',
    details: report.details?.trim() || 'No details were given.',
    reportedBy: report.reportedBy?.trim() || 'Worker',
    reportedAt: report.reportedAt,
    assignedToUid: to.uid,
    assignedToName: to.name,
    assignedToWorkerId: to.workerId,
    assignedAt: now,
    status: 'assigned',
    startedAt: null,
    fixedAt: null,
    fixNote: null,
  };
}

/** One line for the admin's report card. */
export function repairStatusText(job: Pick<RepairJobRecord, 'status' | 'assignedToName'>): string {
  switch (job.status) {
    case 'in_progress':
      return `${job.assignedToName} is repairing it now`;
    case 'fixed':
      return `Fixed by ${job.assignedToName} -- please confirm`;
    default:
      return `Assigned to ${job.assignedToName} -- waiting to start`;
  }
}

/** A "fixed" job the admin still has to confirm (the underlying report isn't closed yet) triggers a notification. */
export function repairAwaitsAdmin(
  job: Pick<RepairJobRecord, 'status' | 'kind' | 'reportId'>,
  reports: { equipment: EquipmentConditionReport[]; sprinkler: IrrigationDamageReportRecord[] },
): boolean {
  if (job.status !== 'fixed') return false;
  if (job.kind === 'equipment') {
    const r = reports.equipment.find((x) => x.reportId === job.reportId);
    return Boolean(r) && !(r!.reviewed || r!.fixedAt);
  }
  const r = reports.sprinkler.find((x) => x.reportId === job.reportId);
  return Boolean(r) && (r!.status ?? 'Pending').trim().toLowerCase() !== 'resolved';
}

File: app/src/main/java/com/melodypenero/coffeefarm/domain/RepairJobs.kt
package com.melodypenero.coffeefarm.domain

import com.melodypenero.coffeefarm.data.store.WorkerRecord

/** Kinds of report the admin can turn into a repair job. Add another here (and on the website) to support more. */
object RepairKind {
    const val EQUIPMENT = "equipment"
    const val SPRINKLER = "sprinkler"
}

/** Progress a Maintenance worker reports on a job. `assigned` is set by the admin; the worker sets the other two. */
object RepairStatus {
    const val ASSIGNED = "assigned"
    const val IN_PROGRESS = "in_progress"
    const val FIXED = "fixed"
}

/** A fixed job needs a note saying what was done; firestore.rules enforces the same 1-500 character limit. */
const val MAX_FIX_NOTE_CHARS = 500

fun isValidFixNote(note: String): Boolean = note.trim().let { it.isNotEmpty() && it.length <= MAX_FIX_NOTE_CHARS }

/** A part or material the worker used up in a repair (e.g. 4 screws); the admin deducts it from the supplies. */
data class PartUsed(val name: String, val quantity: Double)

const val MAX_PARTS_PER_JOB = 10
const val MAX_PART_NAME_CHARS = 60
const val MAX_PART_QUANTITY = 100_000.0

/**
 * Keeps only what a quantity can contain while typing or pasting: digits and one decimal point, at most 2 decimals
 * and 6 whole digits. Letters, symbols, commas, minus signs and a second "." are dropped.
 */
fun sanitizePartQuantity(raw: String): String {
    val cleaned = raw.filter { it in '0'..'9' || it == '.' }
    val dot = cleaned.indexOf('.')
    if (dot < 0) return cleaned.take(6)
    val whole = cleaned.substring(0, dot).take(6)
    val decimals = cleaned.substring(dot + 1).filter { it != '.' }.take(2)
    return "$whole.$decimals"
}

/** The typed quantity, or null if it isn't more than 0 and at most [MAX_PART_QUANTITY]. */
fun parsePartQuantity(text: String): Double? {
    val v = text.trim().takeIf { it.isNotEmpty() && it != "." }?.toDoubleOrNull() ?: return null
    return v.takeIf { it > 0.0 && it <= MAX_PART_QUANTITY }
}

/**
 * Turns the rows typed in the fix dialog (name, quantity) into the parts to send. Fully blank rows are ignored.
 * Returns null while any other row is incomplete or invalid (a name without a valid quantity, or a quantity
 * without a name) -- the dialog keeps "Yes, it's fixed" disabled until that is sorted out.
 */
fun parsePartsUsed(rows: List<Pair<String, String>>): List<PartUsed>? {
    val parts = mutableListOf<PartUsed>()
    for ((rawName, rawQty) in rows) {
        val name = rawName.trim().replace(Regex("\\s+"), " ")
        if (name.isEmpty() && rawQty.isBlank()) continue
        val qty = parsePartQuantity(rawQty)
        if (name.isEmpty() || name.length > MAX_PART_NAME_CHARS || qty == null) return null
        parts += PartUsed(name, qty)
    }
    return if (parts.size <= MAX_PARTS_PER_JOB) parts else null
}

/** "4 × Screw, 1 × Nozzle" -- how a list of parts reads on a card. */
fun partsSummary(parts: List<PartUsed>): String =
    parts.joinToString(", ") { "${formatPartQuantity(it.quantity)} × ${it.name}" }

fun formatPartQuantity(q: Double): String = if (q == Math.floor(q)) q.toLong().toString() else "%.2f".format(java.util.Locale.US, q).trimEnd('0').trimEnd('.')

/**
 * A repair job assigned to the signed-in Maintenance worker (a `repair_jobs` document). It carries a copy of
 * the report -- the worker can't read other workers' private report documents -- so everything they need to
 * know what to fix is here.
 */
data class RepairJob(
    val jobId: String,
    val kind: String,
    val reportId: String,
    /** Equipment name or sprinkler label. */
    val title: String,
    /** e.g. "Broken / wrecked" or "Zone: Zone B". */
    val subtitle: String,
    /** What the reporting worker said was wrong. */
    val details: String,
    val reportedBy: String,
    val reportedAt: String,
    val assignedToName: String,
    val assignedAt: String,
    val status: String,
    val startedAt: String,
    val fixedAt: String,
    val fixNote: String,
    /** Parts / materials the worker reported using; the admin deducts them from the supplies. */
    val partsUsed: List<PartUsed> = emptyList()
) {
    val isFixed: Boolean get() = status == RepairStatus.FIXED
    val kindLabel: String get() = if (kind == RepairKind.SPRINKLER) "Sprinkler" else "Equipment"
}

/** Builds a [RepairJob] from a raw Firestore document map. Tolerant of missing fields. */
fun parseRepairJob(jobId: String, data: Map<String, Any?>): RepairJob {
    fun str(key: String) = data[key]?.toString().orEmpty()
    return RepairJob(
        jobId = jobId,
        kind = str("kind").ifBlank { RepairKind.EQUIPMENT },
        reportId = str("reportId"),
        title = str("title").ifBlank { "Repair job" },
        subtitle = str("subtitle"),
        details = str("details"),
        reportedBy = str("reportedBy"),
        reportedAt = str("reportedAt"),
        assignedToName = str("assignedToName"),
        assignedAt = str("assignedAt"),
        status = str("status").ifBlank { RepairStatus.ASSIGNED },
        startedAt = str("startedAt"),
        fixedAt = str("fixedAt"),
        fixNote = str("fixNote"),
        partsUsed = (data["partsUsed"] as? List<*>).orEmpty().mapNotNull { item ->
            val m = item as? Map<*, *> ?: return@mapNotNull null
            val name = m["name"]?.toString()?.trim().orEmpty()
            val qty = (m["quantity"] as? Number)?.toDouble()
            if (name.isEmpty() || qty == null || qty <= 0.0) null else PartUsed(name, qty)
        }
    )
}

/** Jobs to do first (already-started ones on top, then oldest assignment first); fixed jobs after, newest first. */
fun sortRepairJobs(jobs: List<RepairJob>): List<RepairJob> {
    val (done, todo) = jobs.partition { it.isFixed }
    val ordered = todo.sortedWith(
        compareBy<RepairJob> { if (it.status == RepairStatus.IN_PROGRESS) 0 else 1 }.thenBy { it.assignedAt }
    )
    return ordered + done.sortedByDescending { it.fixedAt.ifBlank { it.assignedAt } }
}

/**
 * The worker record for this login if -- and only if -- they are an active Maintenance worker. Matched by the login's
 * Firebase uid (set when the admin created the worker account) or the account email.
 */
fun findMaintenanceWorker(userId: String, email: String, workers: List<WorkerRecord>): WorkerRecord? =
    workers.firstOrNull { w ->
        w.isActive && FarmFinance.isMaintenanceRole(w.roleRate) &&
            ((w.authUid.isNotBlank() && w.authUid == userId) ||
                (w.accountEmail.isNotBlank() && w.accountEmail.equals(email.trim(), ignoreCase = true)))
    }
